File: leetcode/0020_valid_parentheses.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <stack>
#include <string>

using namespace std;

class Solution {
 public:
    int isValid(string s) {
        stack<char> stack;

        for (auto c : s) {
            if (c == '(') {
                stack.push(')');
            } else if (c == '[') {
                stack.push(']');
            } else if (c == '{') {
                stack.push('}');
            } else if (stack.empty()) {
                return false;
            } else {
                char top = stack.top();
                stack.pop();

                if (top != c) {
                    return false;
                }
            }
        }

        return stack.size() == 0;
    }
};

int main() {
    auto solution = Solution();

    assert(solution.isValid("()"));
    assert(solution.isValid("()[]{}"));
    assert(!solution.isValid("(]"));
    assert(!solution.isValid("([)]"));
    assert(solution.isValid("{[]}"));
    assert(!solution.isValid("]"));
    assert(!solution.isValid("["));
}

File: leetcode/0020_valid_parentheses.py
# -*- coding: utf-8 -*-

from collections import deque


class Solution:
    def isValid(self, s):
        stack = deque()

        for c in s:
            if c == '(':
                stack.append(')')
            elif c == '[':
                stack.append(']')
            elif c == '{':
                stack.append('}')
            elif c == ')' or c == ']' or c == '}':
                try:
                    top = stack.pop()
                    if top != c:
                        return False
                except IndexError:
                    return False

        return len(stack) == 0


if __name__ == '__main__':
    solution = Solution()

    assert solution.isValid('()')
    assert solution.isValid('()[]{}')
    assert not solution.isValid('(]')
    assert not solution.isValid('([)]')
    assert solution.isValid('{[]}')
    assert not solution.isValid(']')
    assert not solution.isValid('[')
